### ISSUE TITLE
HashLocationStrategy should leave URL untouched before hash

### ISSUE THREAD

**Issue by jfrantzius** [angular/angular#20033]
## I'm submitting a...

<pre><code>[x] Feature request</code></pre>

## Current behavior
Currently, ``HashLocationStrategy`` will replace the complete current URL in the browser with the one given as ``<base href=...`` in the embedding HTML. This means it manipulates the URL _before_ the # sign, even though its [documentation](https://angular.io/api/common/HashLocationStrategy#hashlocationstrategy) says: 
>  Use URL **hash** for storing application location data.

## Expected behavior
HashLocationStrategy should manipulate the browser URL only after the # sign in the URL. 

## What is the motivation / use case for changing the behavior?
The current behaviour of ``HashLocationStrategy`` is both against the intuition of its name, as well as not mentioned in its documentation.

But most of all, this is cumbersome in integration scenarios, where an Angular app is embedded in a page e.g. served by a Content Management System. It's cumbersome because setting ``<base href="./"/>`` to the current page will distort asset loading, as that also depends on the base href. There seems to be a workaround for the latter by using ``--deploy-url`` on building the app, but that's already two steps more to find out, and also very badly documented...

**Comment by jasonaden** [comment]
I'm sorry but reported issues require a reproduction of the issue. We suggest using [StackBlitz](https://stackblitz.com/fork/angular-gitter), [Plunker](http://plnkr.co/edit/tpl:AvJOMERrnz94ekVua0u5?p=catalogue) or [Github](https://github.com/new).

If this issue persists, please create a reproduction using one of the links above. Describe the difference between the expected and current behavior in a new issue with the reproduction linked.
